specs/Spec.Agile.AEAD.fsti

```diff
@@ -119,7 +121,7 @@ let decrypted #a (c: cipher a) = p:plain a { S.length c = cipher_length p }
 // specs. Sigh.
 val encrypt: #(a: supported_alg) -> kv a -> iv a -> ad a -> p:plain a -> encrypted p
 val decrypt: #(a: supported_alg) -> kv a -> iv a -> ad a ->
-  c:cipher a { S.length c <= max_length a + tag_length a} ->
+  c:cipher a { S.length c <= cipher_max_length a} -> // TODO If my change of cipher can stay, then this check here is redundant
```

Comment:
I am sorry for having left TODOs.

I want to make a case for keeping this change, i.e., adding the restriction to the `cipher` type (with my most recent commit removing TODO comments and simplifying the signature further). My reasoning in the following.

The type `plain` is already a bytestring with maximum length:

```
let plain (a: supported_alg) = s:S.seq uint8 { S.length s <= max_length a }
```

Thus, the `encrypt` function will only produce `cipher` values with some maximum length. However, the `cipher` type itself is not restricted, prior to this PR:

```
let cipher (a: supported_alg) = s:S.seq uint8 { S.length s >= tag_length a }
```

Furthermore, the `decrypt` function only takes a `cipher` value up to a certain maximum length:

```
val decrypt: #(a: supported_alg) -> kv a -> iv a -> ad a ->
  c:cipher a { S.length c <= max_length a + tag_length a} ->
```

This makes the “usability” of the type `cipher` somewhat limited: it contains bytestrings longer than the maximum allowed by `decrypt`, but these values cannot be produced by `encrypt` and also cannot be processed by `decrypt`.

Furthermore, in HPKE, while we can simply use `plain` as parameter in the outwards facing API of encryption functions:

```
val sealBase:
    cs:ciphersuite_not_export_only
  -> skE:key_dh_secret_s cs
  -> pkR:DH.serialized_point (kem_dh_of_cs cs)
  -> info:info_s cs
  -> aad:AEAD.ad (aead_alg_of cs)
  -> pt:AEAD.plain (aead_alg_of cs) ->
  Tot (option (key_dh_public_s cs & AEAD.encrypted #(aead_alg_of cs) pt))
```

we need to have the additional restriction required by `decrypt` in each function signature that does decryption:

```
val openBase:
    cs:ciphersuite_not_export_only
  -> enc:key_dh_public_s cs
  -> skR:key_dh_secret_s cs
  -> info:info_s cs
  -> aad:AEAD.ad (aead_alg_of cs)
  -> ct:AEAD.cipher (aead_alg_of cs) {Seq.length ct <= AEAD.max_length (aead_alg_of cs) + AEAD.tag_length (aead_alg_of cs)} ->
  Tot (option (AEAD.decrypted #(aead_alg_of cs) ct))
```

This can be done, but it is inconvenient: (1) it is necessary in the HPKE spec and in every spec that uses HPKE (2) it exposes the details of the AEAD inner structure (actual ciphertext plus tag) to the exterior.

By the way, in the comments of Spec.Agile.AEAD.fsti, the URL https://github.com/mitls/mitls-papers/wiki/Discussion-to-Converge-on-Design-Decisions is mentioned when talking about length restrictions of `plain`. This URL does not seem to be accessible anymore. GitHub throws an “Upgrade to GitHub Team or make this repository public to enable this feature.” error.